test/e2e/storage/vsphere_statefulsets.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+/*
+	Test performs following operations
+
+	Steps
+	1. Create a storage class with thin diskformat.
+	2. Create nginx service.
+	3. Create nginx statefulsets with 3 replicas.
+	4. Wait until all Pods are ready and PVCs are bounded with PV.
+	5. Verify volumes are accessible in all statefulsets pods with creating empty file.
+	6. Scale down statefulsets to 2 replicas.
+	7. Scale up statefulsets to 4 replicas.
+	8. Scale down statefulsets to 0 replicas and delete all pods.
+	9. Delete all PVCs from the test namespace.
+	10. Delete the storage class.
+*/
+
+const (
+	manifestPath     = "test/e2e/testing-manifests/statefulset/nginx"
+	mountPath        = "/usr/share/nginx/html"
+	storageclassname = "nginx-sc"
+)
+
+var _ = SIGDescribe("vsphere statefulsets", func() {
+	f := framework.NewDefaultFramework("vsphere-statefulsets")
+	var (
+		namespace string
+		client    clientset.Interface
+	)
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("vsphere")
+		namespace = f.Namespace.Name
+		client = f.ClientSet
+	})
+	AfterEach(func() {
+		framework.Logf("Deleting all statefulset in namespace: %v", namespace)
+		framework.DeleteAllStatefulSets(client, namespace)
+	})
+
+	It("vsphere statefulsets testing", func() {
+		By("Creating StorageClass for Statefulsets")
+		scParameters := make(map[string]string)
+		scParameters["diskformat"] = "thin"
+		scSpec := getVSphereStorageClassSpec(storageclassname, scParameters)
+		sc, err := client.StorageV1().StorageClasses().Create(scSpec)
+		Expect(err).NotTo(HaveOccurred())
+		defer client.StorageV1().StorageClasses().Delete(sc.Name, nil)
+
+		By("Creating statefulsets with number of Replica: 3")
+		statefulsetsTester := framework.NewStatefulSetTester(client)
+		statefulsets := statefulsetsTester.CreateStatefulSet(manifestPath, namespace)
+		// Waiting for pods status to be Ready
+		ssPods := statefulsetsTester.GetPodList(statefulsets)
+		Expect(ssPods.Items).NotTo(BeEmpty(), fmt.Sprintf("Unable to get list of Pods from the Statefulset: %v", statefulsets.Name))
```

Comment:
Please move this to right before where it's used, and after `WaitForStatusReadyReplicas` is done. Otherwise, you're checking a stale list of pods which may not contain all of the StatefulSet pods. You need to check the `len` of the podlist as well.

Reply:
Added len check.
```
Expect(len(ssPods.Items) == int(replicas)).To(BeTrue(), "Number of Pods in the statefulset should match with number of replicas")

```
Keeping the Pod retrieval before scaling down statefulset, so that after scale down, I can verify for deleted pods, associated PV is detached from the node.